### PR TITLE
[FLINK-19787][table-runtime] Migrate Filesystem connector to new table source sink interface

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Migrate Filesystem source to new table source sink interface

## Brief change log

- Modify `FileSystemTableFactory`
- Introduce `AbstractFileSystemTable` for some common fields
- Modify `FileSystemTableSource`
- Modify `FileSystemTableSink`

## Verifying this change

This change is already covered by existing tests, such as `FileSystemITCaseBase`.

Need Factory testing for option validation.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no

### CODE REVIEW

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableFactory.java [hunk 1]
```diff
@@ -48,71 +36,59 @@
...
-		return new FileSystemTableSink(
-				context.getObjectIdentifier(),
-				context.isBounded(),
-				TableSchemaUtils.getPhysicalSchema(context.getTable().getSchema()),
-				getPath(conf),
-				context.getTable().getPartitionKeys(),
-				conf.get(PARTITION_DEFAULT_NAME),
-				context.getTable().getOptions());
+	private void validate(FactoryUtil.TableFactoryHelper helper) {
+		// Except format options, some format like parquet and orc can not list all support options.
```
Comment (leonardBang) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableFactory.java:64]:
```suggestion
		// Except format options, some formats like parquet and orc can not list all supported options.
```

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSource.java [hunk 2]
```diff
@@ -226,87 +195,63 @@ public long getPushedDownLimit() {
...
-				path,
-				partitionKeys,
-				defaultPartName,
-				properties,
-				readPartitions,
-				selectFields,
-				limit,
-				new ArrayList<>(predicates));
+	public String asSummaryString() {
+		return "filesystem";
```
Comment (leonardBang) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSource.java:226]: Filesystem

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSink.java [hunk 3]
```diff
@@ -351,19 +315,34 @@ public void setStaticPartition(Map<String, String> partitions) {
...
-	public boolean configurePartitionGrouping(boolean supportsGrouping) {
-		this.dynamicGrouping = supportsGrouping;
-		return dynamicGrouping;
+	public DynamicTableSink copy() {
+		return new FileSystemTableSink(context, overwrite, dynamicGrouping, staticPartitions);
+	}
+
+	@Override
+	public String asSummaryString() {
+		return "filesystem";
```
Comment (leonardBang) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/FileSystemTableSink.java:335]: FileSystem

**Review by leonardBang — COMMENTED** [review]
@JingsongLi thanks for the contribution, I left few comments

**Review by leonardBang — APPROVED** [review]
Thanks @JingsongLi for the update, LGTM once the azure pipeline passed

**Comment by JingsongLi** [comment]
Thanks @leonardBang for the review, merging...

**Pull request merged**